<a href="https://colab.research.google.com/github/oluarolu/checkfolder/blob/master/Into_to_pytorch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import torch library

In [0]:
import torch

initialize a tensor

In [0]:
a = torch.ones((2, 2), requires_grad=True)

a

tensor([[1., 1.],
        [1., 1.]], requires_grad=True)

perform operations on the tensor

In [0]:
b = a + 5

In [0]:
c = b.mean()

In [0]:
print(b, c)

tensor([[6., 6.],
        [6., 6.]], grad_fn=<AddBackward0>) tensor(6., grad_fn=<MeanBackward0>)


The derivative of c wrt to a will be 1/4 hence the gradient of the matric will be 0.25. Checking with pytourch

back propagating

In [0]:
c.backward()

computin the gradient

In [0]:
print(a.grad)

tensor([[0.5000, 0.5000],
        [0.5000, 0.5000]])


import Optim module -pytorch modle for optimizers

In [0]:
from torch import optim

Implementing a Neural Network from Scracth

Initialize the input tensor

In [0]:
X = torch.Tensor([[1, 0, 1, 0], [1, 0, 1, 1], [0, 1, 0, 1]])

X

tensor([[1., 0., 1., 0.],
        [1., 0., 1., 1.],
        [0., 1., 0., 1.]])

Output

In [14]:
y =torch.Tensor([[1], [1], [0]])

y

tensor([[1.],
        [1.],
        [0.]])

In [15]:
print(X, '\n')

print(y)

tensor([[1., 0., 1., 0.],
        [1., 0., 1., 1.],
        [0., 1., 0., 1.]]) 

tensor([[1.],
        [1.],
        [0.]])


define as sigmoid function that will act as activation function

define a sigmoid function

In [0]:
def sigmoid (x):
  return 1/(1 + torch.exp(-x))

define derivative of Sigmoid function

In [0]:
def derivatives_sigmoid(x):
  return sigmoid(x) * (1 - sigmoid(x))

initialize parameters for the model including epoch, learning rate, weights, biases etc

initializing the variables

initailizing the epoch

In [0]:
epoch = 7000 #Setting training iterations

In [0]:
lr = 0.1 # Setting learning rate

In [0]:
inputlayer_neurons = X.shape[1] #number of features in data set

In [0]:
hiddenlayer_neurons =3

In [0]:
output_neurons =1 #no of neurons in output layer

Weight and biase initialization

In [0]:
wh =torch.randn(inputlayer_neurons, hiddenlayer_neurons).type(torch.FloatTensor)

In [0]:
bh =torch.randn(1, hiddenlayer_neurons).type(torch.FloatTensor)

In [0]:
wout = torch.randn(hiddenlayer_neurons, output_neurons)

In [0]:
bout = torch.randn(1, output_neurons)

creating the neural network

In [0]:
for i in range(epoch):
  #forward propagation
  hidden_layer_input1 = torch.mm(X, wh)
  hidden_layer_input = hidden_layer_input1 + bh
  hidden_layer_activations = sigmoid(hidden_layer_input)

  output_layer_input1 = torch.mm(hidden_layer_activations, wout)
  output_layer_input = output_layer_input1 + bout
  output = sigmoid(output_layer_input)

  #backpropagation
  E = y - output
  slope_output_layer = derivatives_sigmoid(output)
  slope_hidden_layer = derivatives_sigmoid(hidden_layer_activations)
  d_output = E * slope_output_layer
  Error_at_hidden_layer = torch.mm(d_output, wout.t())
  d_hiddenlayer = Error_at_hidden_layer * slope_hidden_layer


  wout += torch.mm(hidden_layer_activations.t(), d_output) * lr
  bout += d_output.sum() * lr
  wh += torch.mm(X.t(), d_hiddenlayer) * lr
  bh += d_output.sum() * lr

check the output of the model

In [46]:
print('actual :\n', y, '\n')
print('predicted :\n', output)

actual :
 tensor([[1.],
        [1.],
        [0.]]) 

predicted :
 tensor([[0.9978],
        [0.9974],
        [0.0044]])


The target is 1, 1, 0 and the predicted values from the model are 0.9978, 0.9974, 0.0044